subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cpu.h

```diff
@@ -16,12 +16,21 @@ static inline void cpu_sleep(void)
 #endif
 }
 
-static inline void cpu_dsb(void)
+static inline void cpu_dmb(void)
 {
-#if defined(CONFIG_CPU_CORTEX_M0) || defined(CONFIG_ARCH_POSIX)
-	/* No need of data synchronization barrier */
-#elif defined(CONFIG_CPU_CORTEX_M4) || defined(CONFIG_CPU_CORTEX_M33)
-	__DSB();
+#if defined(CONFIG_CPU_CORTEX_M0) || \
+	defined(CONFIG_CPU_CORTEX_M4) || \
+	defined(CONFIG_CPU_CORTEX_M33)
+	/* NOTE: __DMB() shall have a compiler memory barrier (clobber)
+	 *      aka __asm__ volatile ("dmb sy" : : : "memory");
+	 */
+	__DMB();
```

Comment:
IF the problem is only the re-ordering, why use a data memory barrier? _DMB() is really not needed in Cortex-M. Instruction re-ordering could be avoided with a compiler barrier, or am I missing something?

Reply:
Based on ARM recommendation:
> On microcontroller devices using the Cortex-M3 and Cortex-M4 processors, omission of the DMB instruction in semaphore and Mutex operations does not cause an error. However, this is not guaranteed if:
•the processor has a cache
•the software is used in a multi-core system.
ARM recommends that the DMB instruction should be used in semaphores and Mutex operations in OS designs.

will keep the DMB instruction to be safe with respect to if cache is implemented in processor.

Reply:
but that certainly does not apply to Cortex-M0, for example, and even M4, I am not sure.
My suggestion is to limit the solution to the minimal: Instruction re-orderning -> compliler barrier, so we are not hiding with DSB/DMB any other corner cases that we can not later detect and educate ourselves.

Reply:
As long as the Zephyr kernel code is using the recommendations, I will be against taking short cuts. Let zephyr kernel code, remove the use of DSB/DMB first. This is something yourself added in some places in the ARM architecture code.

Reply:
DMB should not really be added anywhere, if i remember correctly. If it is, it should be removed, agreed. As for DSB (and ISB), this is added, for reasons other than fighting instruction re-ordering. See relevant comments below :) .